### PR TITLE
Only fetch one collation at a time per relay parent

### PR DESCRIPTION
**Pull request by bkchr**
Before a validator would fetch all collations that were advertised to
him. This pr changes the behavior to always just fetch one collation at
a time. If fetching fails, the validator will start fetching one of the
other collations.

### CODE REVIEW

File: node/network/collator-protocol/src/validator_side/tests.rs [hunk 1]
```diff
@@ -256,6 +256,135 @@ async fn respond_to_core_info_queries(
 	);
 }
 
+/// Assert that the next message is a `CandidateBacking(Second())`.
+async fn assert_candidate_backing_second(
```
Comment (ordian) [inline node/network/collator-protocol/src/validator_side/tests.rs:260]: :+1:

**Review by ordian — APPROVED** [review]
Looks good.

**Comment by ordian** [comment]
please merge master

File: node/network/collator-protocol/src/validator_side/mod.rs [hunk 2]
```diff
@@ -1119,6 +1111,67 @@ pub(crate) async fn run<Context>(
...
+	let (candidate_receipt, pov) = match res {
+		Ok(res) => res,
+		Err(_) => {
+			tracing::debug!(
+				target: LOG_TARGET,
+				relay_parent = ?collation_event.1.relay_parent,
+				para_id = ?collation_event.1.para_id,
+				peer_id = ?collation_event.1.peer_id,
+				collator_id = ?collation_event.0,
+				"Collation fetching has timed out.",
```
Comment (pepyakin) [inline node/network/collator-protocol/src/validator_side/mod.rs:1134]: Is this only a time out? AFAIU a sudden disconnection will lead to the same result and thus incorrectly be reported as a time out.

File: node/network/collator-protocol/src/validator_side/mod.rs [hunk 3]
```diff
@@ -1119,6 +1111,67 @@ pub(crate) async fn run<Context>(
 	Ok(())
 }
 
+/// Handle a fetched collation.
+async fn handle_collation_fetched(
```
Comment (pepyakin) [inline node/network/collator-protocol/src/validator_side/mod.rs:1115]: Nit: It's not that we call this function when collation is fetched, but rather when collation fetch result is received. A name and rustdoc update would be good.

File: node/network/collator-protocol/src/validator_side/mod.rs [hunk 4]
```diff
@@ -770,22 +802,23 @@ where
...
+					let collations = state.collations_per_relay_parent.entry(relay_parent).or_default();
+
+					if matches!(collations.status, CollationStatus::Fetching) {
+						collations.unfetched_collations.push((pending_collation, id));
+					} else if matches!(collations.status, CollationStatus::Waiting) {
+						collations.status = CollationStatus::Fetching;
+						drop(collations);
+
+						fetch_collation(ctx, state, pending_collation.clone(), id).await;
+					}
```
Comment (pepyakin) [inline node/network/collator-protocol/src/validator_side/mod.rs:821]: Nit: that sounds like a job for a `match` )

File: node/network/collator-protocol/src/validator_side/mod.rs [hunk 5]
```diff
@@ -982,12 +1021,10 @@ where
...
 		}
 		Invalid(parent, candidate_receipt) => {
-			if match state.pending_candidates.get(&parent) {
-				Some(collation_event)
-					if Some(candidate_receipt.commitments_hash) == collation_event.1.commitments_hash
-				=> true,
-				_ => false,
-			} {
+			if state.pending_candidates
+				.get(&parent).map(|e| e.1.commitments_hash == Some(candidate_receipt.commitments_hash))
```
Comment (pepyakin) [inline node/network/collator-protocol/src/validator_side/mod.rs:1025]: Nit: could use a newline

**Review by pepyakin — APPROVED** [review]
All nits, feel free to ignore

**Review by rphmeier — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**